### PR TITLE
Render smaller fields on top of larger ones

### PR DESCRIPTION
**Pull request by Blue-Award**
Per discussion at https://freeorion.org/forum/viewtopic.php?p=116994#p116994 smaller fields could "hide" under bigger ones and it could be hard to fish them out with a mouse.

https://github.com/freeorion/freeorion/pull/4739 improved functional aspect of that, putting the fields in z-order according to their size, so as far as mouse events are concerned, smaller fields are on top of larger ones.

However, I've uncovered that visual aspect of this is separate. Field icons don't render themselves, they are rendered by galaxy map window. And this is separate processing, grouped by field texture. So e.g. all ion storms would be rendered before molecular clouds, or vice versa (depending on somewhat random factors).

So, a smaller field could still get rendered visually under a bigger one, and moreover, due to grouping by texture, it's not even possible to avoid this situation in general case as is (a smaller ion storm could be rendered on top of larger molecular cloud, but in other place in the galaxy, smaller molecular cloud would be rendered under bigger ion storm... or vice versa)

Seems to me the grouping by texture is not necessary, even if it ends up with fewer GL calls. I propose to render fields more individually, largest to smallest, so smaller ones get painted over larger ones. While still grouping by texture as much as possible.

Following picture shows situation from MP25, where an ion storm is between two molecular clouds... Ion storm is on top of larger molecular cloud, but the smaller molecular cloud is visually rendered under ion storm (it could also happen that the ion storm is rendered under both the larger molecular cloud and smaller molecular cloud).

After this change, all three fields in this configuration are rendered per their size, matching the improved mouse behavior

spot_the_difference

### CODE REVIEW

**Comment by Blue-Award** [comment]
So I went with some std::views things, but turns out something like "ubuntu gcc" builds but "ubuntu clang" doesn't. Seems I need to use std::ranges::views instead, or somehow declare myself that std::views is std::ranges::views? 

There's also util/ranges.h thing and I think I could use range_filter from there, though because of the above I am confused if in clang's case here, would it actually use std::views::filter, or the boost::adaptors::filtered thing? 

So at least std::ranges::views::filter I could cover with range_filter (and ignore the thing about clang possibly using boost adaptor) But someting like std::views::reverse or std::views::join? I'll change the stuff to std::ranges::views from std::views and see how that builds for macos and clang

Would you recon util/ranges.h could use some update?

**Comment by Blue-Award** [comment]
Hmm is there any way to access std::views::reverse and std::views::join in macos and clang builds? I understand util/ranges.h gives some workarounds for std::views::filter and some others, but it looks to give up hope on "join" and doesn't include reverse.

Clang build seemed to tell me it doesn't know what std::views::xxx is, but seemed to find some stuff under std::ranges::views::xxx but when I explicitly put std::ranges::views it doesn't seem to be happy. Maybe it actually has <ranges> nowadays, but is not included due to trying some workarounds in ranges.h?

Well I think I'll use good old reverse iterator and maybe for_each instead of views::reversed, and drop using views::join, instead making two loops or something. Wouldn't it be nice though if those worked for all the builds, eh?

**Comment by geoffthemedio** [comment]
> There's also util/ranges.h thing and I think I could use range_filter from there, though because of the above I am confused if in clang's case here, would it actually use std::views::filter, or the boost::adaptors::filtered thing?

It will use std ranges if available and functional as best I could determine, and otherwise fall back to boost.

> std::views::reverse or std::views::join?

I didn't yet need to reverse a view, so haven't tried to add it to ranges.h. If you do and there is something equivalent in std and Boost, feel free to add it. But just using `rbegin` would probably work.

Boost has no equivalent for `std::views::join` that I could find:  https://github.com/freeorion/freeorion/blob/master/util/ranges.h#L11

**Comment by Blue-Award** [comment]
Generally, does this mean clang and macos builds are not on C++20? Again, I had impression clang knew about std::ranges::views when I tried std::views, that's why I switched to that namespace, but maybe something else was missing

The great interwebs often suggest using https://github.com/ericniebler/range-v3 which seems to be the basis of range additions to C++20. Is that something we could pull into sdk to fall back onto instead of boost?

Pushing children down the stack works one way (push smaller first so they end on top) and drawing works the opposite way (render larger first so smaller end on top) hence reversal is handy in one or the other place but yea if I don't figure the view thing I will resort to rbegin

Joining and views::all just to flatten two vectors and go over them in one loop instead of nested loop or two loops. I had to explicitly do views::all otherwise I was ending up with const references and didn't want to do const_cast. In this case I think I could use boost::join but it is not a replacement to views that flatten many ranges not just two

**Comment by geoffthemedio** [comment]
> Generally, does this mean clang and macos builds are not on C++20? Again, I had impression clang knew about std::ranges::views when I tried std::views, that's why I switched to that namespace, but maybe something else was missing

I don't remember all the details, but compilers rarely support everything from any given standard, and different compilers support different subsets, and different versions of those compilers will support different amounts, and we officially support not always the latest versions of those compilers.

> The great interwebs often suggest using https://github.com/ericniebler/range-v3 which seems to be the basis of range additions to C++20. Is that something we could pull into sdk to fall back onto instead of boost?

Possibly, but I don't think we'll add another dependency for that.

> Joining and views::all just to flatten two vectors and go over them in one loop instead of nested loop or two loops. I had to explicitly do views::all otherwise I was ending up with const references and didn't want to do const_cast.

Do not `const_cast` to avoid such compiler issues.

> In this case I think I could use boost::join but it is not a replacement to views that flatten many ranges not just two

There is a distinction between flattening a range of ranges and concatenating two unrelated ranges into one, and std and Boost choose differently for their respective `join` views.

**Comment by Blue-Award** [comment]
Yeah didn't want to use const_cast and what helped in my particular case was views::all but could be solved many other ways

There is boost::adaptors::reversed https://www.boost.org/doc/libs/1_83_0/libs/range/doc/html/range/reference/adaptors/reference/reversed.html so I'll see into adding that to https://github.com/freeorion/freeorion/blob/master/util/ranges.h

**Comment by Blue-Award** [comment]
to avoid std::views::join, I went with two loops. Originally, to avoid const reference I was getting from using initialization list, I went with pointers that didn't present me with const reference problem:

`for (auto* field_vertices : { &m_field_vertices_not_visible, &m_field_vertices_visible }) {
        for (auto& [field_texture, buffer] : (*field_vertices) | range_filter(field_texture_exists)) {`

but then I see I could also go with the following (reference_wrappers that also unwrap without const):

`for (auto& field_vertices : { std::ref(m_field_vertices_not_visible), std::ref(m_field_vertices_visible) }) {
        for (auto& [field_texture, buffer] : (field_vertices.get()) | range_filter(field_texture_exists)) {`

kind of po-tay-to, po-tah-to, no? But still, the latter seems nicer to me? So I'll update with that. But maybe there could be even nicer/more kosher way to do this, I wonder... splitting hair a bit, perhaps

**Comment by Blue-Award** [comment]
The Internets do have some suggestions, if you focus on "flattening range of ranges" rather than "joining ranges". Should be possible to have a reasonable replacement for std::views::join in terms of boost, though not sure how badly needed that would be (if it were really badly needed then prolly importing range-v3 stuff would be more desirable...). 

Still, that's story for another day I would think. I am happy with doing loop in a loop here to avoid code duplication in the particular bit. Still sounds better to me than e.g. making lambda to share between two for_each loops. But eh could go with that too.

**Comment by Blue-Award** [comment]
I suppose I can't "nominate a reviewer" if not part of (triage) team?

File: UI/MapWnd.cpp [hunk 1]
```diff
@@ -4119,24 +4118,26 @@ void MapWnd::InitFieldRenderingBuffers() {
...
-
-        // TODO: why the binding here?
-        glBindTexture(GL_TEXTURE_2D, field_texture->OpenGLId());
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
-        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);
+    std::size_t max_buffer_size = 0;
 
-        buffers.first.createServerBuffer();
-        buffers.second.createServerBuffer();
+    auto field_texture_exists = [](const auto& field_vertices_pair) { return !!field_vertices_pair.first; };
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: bit odd to write `!!` rather than make the lambda return bool with `-> bool`

Comment (geoffthemedio) [inline UI/MapWnd.cpp]: lambdas that capture nothing can generally be `static constexpr`. those that do capture something can generally be explicitly made `const`

File: UI/MapWnd.cpp [hunk 2]
```diff
@@ -4119,24 +4118,26 @@ void MapWnd::InitFieldRenderingBuffers() {
...
+    for (auto& field_vertices : { std::ref(m_field_vertices_not_visible), std::ref(m_field_vertices_visible) }) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:4126]: can these be `std::cref` ?

File: UI/MapWnd.cpp [hunk 3]
```diff
@@ -4065,9 +4060,13 @@ void MapWnd::InitFieldRenderingBuffers() {
         if (!field_texture)
             continue;
 
-        auto& field_both_vertex_buffers = m_field_vertices[field_texture];
+        // group by texture as much as possible for fewer GL calls, but generally paint fields one by one according to size
+        // so smaller ones get painted over larger ones, including across different textures
+        auto& field_vertices = current_field_visible ? m_field_vertices_visible : m_field_vertices_not_visible;
         GG::GL2DVertexBuffer& current_field_vertex_buffer =
-            current_field_visible ? field_both_vertex_buffers.first : field_both_vertex_buffers.second;
+            (field_vertices.empty() || field_vertices.back().first != field_texture) ?
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: this is a bit hard to parse...

File: UI/MapWnd.cpp [hunk 4]
```diff
@@ -1859,15 +1859,13 @@ void MapWnd::RenderFields() {
     glDisableClientState(GL_COLOR_ARRAY);
     glEnableClientState(GL_TEXTURE_COORD_ARRAY);
 
+    auto buffer_not_empty = [](const auto& field_buffer) { return !field_buffer.second.empty(); };
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: can probably be `static constexpr`

File: UI/MapWnd.cpp (hunk 1 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: I seemed to recall that "back in my day" it was considered safer to do double bang than relying on operator bool. Tried to find something on it, search came with https://stackoverflow.com/questions/3193201/how-does-shared-ptr-safely-allow-casting-to-bool and talks about safe bool idiom and how such stuff may no longer be necessary in post C++11 world. Well shared_ptr seems to be in std since then, too... Before you wrote your own stuff or had boost's shared_ptr (which probably was doing the safe bool idiom anyway so perhaps was okay to relay on operator bool, too, in boost's case?) 

Might be some ancient superstition, doesn't look that weird to me, but I'll be hip and down with the cool kids or something,, hah!! :) I'll change it!!

File: UI/MapWnd.cpp (hunk 2 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp:4126]: no, it's the problem I was referring to earlier, that I get const reference to something I need to change and don't want to do `const_cast`. Specifically `buffer.createServerBuffer();` needs the buffer not to be `const`

File: UI/MapWnd.cpp (hunk 3 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: umm should I put some comment on it then or...? I kept it referring to ``GG::GL2DVertexBuffer& current_field_vertex_buffer `` so there are fewer changes between versions, showing parts did not change. 

The condition means if the buffer collection is empty, or last "buffered" texture is different than the current one, we need to create new vertex buffer for current texture and add new vertices to that, otherwise we can keep adding to the old vertex buffer, because texture didn't change. This groups the gl calls per texture as much as possible while still outputting fields in order of size accross textures

File: UI/MapWnd.cpp (hunk 2 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp:4126]: note this is the place I could use `boost::range::join` because it luckily involves only two ranges (while originally I wanted `std::views::join`) but decided against it (and against adding boost view/adaptor replacement for `std::views::join`)

File: UI/MapWnd.cpp (hunk 1 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: Cheers, make sense, thank you, will change this

File: UI/MapWnd.cpp (hunk 3 above)
Comment (geoffthemedio) [reply UI/MapWnd.cpp]: maybe just put the `(field_vertices.empty() || field_vertices.back().first != field_texture)` into a named variable instead of in the same big expression as the `emplace` call?

File: UI/MapWnd.cpp (hunk 1 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: whelp, I tried `->bool` and it complains 

Error (active)	E0413	no suitable conversion function from "const std::shared_ptr<GG::Texture>" to "bool" exists	

I can however explicitly use `operator bool` though personally would rather do double bang indeed. Sigh. Will do `static constexpr auto field_texture_exists = [](const auto& field_vertices_pair) { return (bool)field_vertices_pair.first; };`

Comment (Blue-Award) [reply UI/MapWnd.cpp]: btw that check seems too defensive to me anyway, is there really a chance the texture would be `nullptr`? but the original code had it so I kept it

File: UI/MapWnd.cpp (hunk 4 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: Actually thinking about it, now that I explicitly split the "visible" and "sensor ghost/not visible" fields into two separate collections, this check seems not necessary. The buffer should never be empty, as now, when it gets created, we do put vertices of at least one field into it

I kept the check from previous code, but it was necessary then, because texture had two buffers assigned in one collection at the same time, never mind if there were any visible/not visible fields of given texture. So you could end up with empty buffer here or there

File: UI/MapWnd.cpp (hunk 1 above)
Comment (geoffthemedio) [reply UI/MapWnd.cpp]: another option is probably `return field_vertices_pair.first.get();`

Comment (Blue-Award) [reply UI/MapWnd.cpp]: OK I'll do get(), that's what the std library implementation for MSVC does anyway under the hood for `(bool)`

Comment (Blue-Award) [reply UI/MapWnd.cpp]: (though it compares to `nullptr` rather than doing pointer casting to bool or whatever)

File: UI/MapWnd.cpp (hunk 3 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: ok, added a further comment and a named bool `should_create_new_buffer`

File: UI/MapWnd.cpp (hunk 1 above)
Comment (geoffthemedio) [reply UI/MapWnd.cpp]: should be able to just return whatever.get() which should be implicitly casted to bool if that's the return type of the lambda. the problem without .get() is that the operator bool() of shared_ptr is explicit, but raw pointer casts to bool are automatic and the result is the same as != nullptr (as far as I know).

File: UI/MapWnd.cpp (hunk 4 above)
Comment (Blue-Award) [reply UI/MapWnd.cpp]: (I removed this check)

**Comment by geoffthemedio** [comment]
@Blue-Award feel free to add yourself to https://github.com/freeorion/freeorion/blob/master/default/credits.xml#L76 or optionally give me a real name to credit with

**Comment by Blue-Award** [comment]
> @Blue-Award feel free to add yourself to https://github.com/freeorion/freeorion/blob/master/default/credits.xml#L76 or optionally give me a real name to credit with

oh thank you! let me think about it, maybe I'd prefer semi-anonimity huh... wouldn't mind getting in the table of monster names or something though :D
